homeassistant/components/kodi/__init__.py

```diff
@@ -48,13 +49,19 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
         entry.data[CONF_SSL],
         session=async_get_clientsession(hass),
     )
+
+    kodi = Kodi(conn)
+
     try:
         await conn.connect()
-        kodi = Kodi(conn)
-        await kodi.ping()
         raw_version = (await kodi.get_application_properties(["version"]))["version"]
+        version = f"{raw_version['major']}.{raw_version['minor']}"
     except CannotConnectError as error:
-        raise ConfigEntryNotReady from error
+        dev_reg = await device_registry.async_get_registry(hass)
+        device = dev_reg.async_get_device({(DOMAIN, entry.entry_id)}, [])
+        if not device:
+            raise ConfigEntryNotReady from error
+        version = device.sw_version
```

Comment:
I'm not sure that's needed - maybe we can simply not include the version in the device info in this situation? I think that unless it's overwritten, it should keep its previous value.

Reply:
`version` needs to be defined for use at https://github.com/home-assistant/core/blob/c98875ed8ec4b03d66344f44c0ba25d4aed404b8/homeassistant/components/kodi/__init__.py#L78-L83
Using the value from the device registry provides us with the last known version.

Reply:
Yes, I'm suggesting we make it optional (pass `None` here, and not use it in `media_player.py` if it's `None`).

Reply:
I agree that it would be a good idea to revise the way the version is handled, but at this point it is assumed everywhere that the version is always available. Maybe this should be fixed in a separate PR?

Reply:
I'm not sure it makes sense to perform the extra work of getting it from the device registry, if we intend to drop it. IIRC there's only one place it's used - `device_info`, and if we make it optional there, we're good.

Reply:
`device_info` is read only when the entity is added to HA and does not seem to be updated afterwards. If the version is not provided at that time, it will not be included in the device information. 
The call to the device registry is relatively inexpensive and makes sure we will only add `media_player` entities for known devices, so even without extracting the version information we should still access it.

As I noted earlier, it would be a good idea to have a proper look at the way the version is handled, but it is in no way related to the issue we're looking at in this PR and should be handled separately.

Reply:
> version is not provided at that time, it will not be included in the device information.

It will always be included when the first time the integration is added. If it's not there in subsequent calls, that's fine, as long as it's not explicitly overwritten.

> As I noted earlier, it would be a good idea to have a proper look at the way the version is handled, but it is in no way related to the issue we're looking at in this PR and should be handled separately.

There's nothing to look at - it's only used in one place, and it's this PR that adds the unneeded complexity. No point in adding all of this, just to remove it in the next PR, when making version optional is simpler to begin with.

> is there a possible scenario where ConfigEntryNotReady is raised with the code in this PR? If not, then we may as well pass on a CannotConnectError if we don't need the version

The only scenario would be if the device was manually deleted, which under normal circumstances would simply recreate the device, but with this PR it will fail. If `version` if optional, we can indeed simply set it to `None` and pass. All that's needed is to make sure in `device_info` to not add it to the dict if it's `None`.

Reply:
OK, so we can `pass` on a `CannotConnectError`.

The problem with making the version info optional is that it will not be updated if HA always starts while kodi is off. It's not just a matter of making it optional, but also a matter of updating the info when required. Granted, the PR as it is now does not cover this scenario either, so we'll have to sort this out either way.

I am working on removing the version information from `__init__.py` and from `device_info` entirely, and instead update the device registry manually in `_on_ws_connected()`. This way, the version information in the device registry gets updated every time we connect to kodi, regardless of the state it had when HA started.

Reply:
The case of a version change after which kodi is always off when the entry is loaded isn't handled by this PR, and is ok to leave to the future. The version isn't used by anything, it's just for display purposes.

Reply:
Fixed in 5d98ded, we move all version related stuff to `_on_ws_connected()`